### PR TITLE
rate limit: move HTTP rate limit config into route table

### PR DESCRIPTION
**Pull request by ccaraman**
This change does a few things
- Removes actions from the filter and moves it to new rate limit configuration(`rate_limits`) under route table.
- Adds support for stages in filters and rate limit configurations. 

This is being done to move to a new HTTP rate limit configurations as specified here https://github.com/lyft/envoy/issues/179

### CODE REVIEW

File: include/envoy/ratelimit/ratelimit.h [hunk 1]
```diff
@@ -90,4 +92,19 @@ class ClientFactory {
 
 typedef std::unique_ptr<ClientFactory> ClientFactoryPtr;
 
+/**
+ * An interface for the rate limit filter configuration.
+ */
+class FilterConfig {
```
Comment (mattklein123) [inline include/envoy/ratelimit/ratelimit.h]: Revert this change.

File: include/envoy/router/ratelimit_router.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  /**
+   * @return the route key, if it exists.
+   */
+  virtual const std::string& routeKey() const PURE;
+
+  /**
+   * @param the stage value to use for comparison finding the set of applicable rate limits.
+   * @return set of RateLimitPolicyEntry that are applicable for a stage.
+   */
+  virtual std::vector<std::reference_wrapper<RateLimitPolicyEntry>>
```
Comment (mattklein123) [inline include/envoy/router/ratelimit_router.h]: const std::vector<std::reference_wrapper<RateLimitPolicyEntry>>&

File: source/common/router/config_impl.cc [hunk 3]
```diff
@@ -15,6 +16,85 @@
 
 namespace Router {
 
+void ServiceToServiceAction::populateDescriptors(const Router::RouteEntry& route,
```
Comment (mattklein123) [inline source/common/router/config_impl.cc]: move all RL stuff into new file (router_ratelimit_impl) or whaever

File: include/envoy/ratelimit/ratelimit.h (hunk 1 above)
Comment (ccaraman) [inline include/envoy/ratelimit/ratelimit.h]: reverted

File: include/envoy/router/ratelimit_router.h (hunk 2 above)
Comment (mattklein123) [inline include/envoy/router/ratelimit_router.h]: If reference_wrapper is giving you issues do:
const std::vector<const RateLimitPolicyEntry*>&

File: include/envoy/router/router_ratelimit.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+namespace Router {
+/**
+* Base interface for reneric rate limit action.
+*/
+class Action {
+public:
+  virtual ~Action() {}
+
+  /**
+   * Potentially populate the descriptor array with new descriptors to query.
```
Comment (mattklein123) [inline include/envoy/router/router_ratelimit.h]: Comment out of date

File: include/envoy/router/router_ratelimit.h [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+   * @param route supplies the target route for the request.
+   * @param descriptors supplies the descriptor array to optionally fill.
+   * @param config supplies the filter configuration.
+   */
+  virtual void populateDescriptors(const RouteEntry& route,
```
Comment (mattklein123) [inline include/envoy/router/router_ratelimit.h]: populateDescriptors is shared with the policy below so should be in a base class.

File: include/envoy/router/router_ratelimit.h [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                                   std::vector<::RateLimit::Descriptor>& descriptors,
+                                   const std::string& local_service_cluster,
+                                   const Http::HeaderMap& headers,
+                                   Http::StreamDecoderFilterCallbacks& callbacks) const PURE;
+};
+
+/**
+* Rate limiting policy
```
Comment (mattklein123) [inline include/envoy/router/router_ratelimit.h]: indent

File: include/envoy/router/router_ratelimit.h [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  /**
+   * @return the route key, if it exists.
+   */
+  virtual const std::string& routeKey() const PURE;
+
+  /**
+   * @param the stage value to use for comparison finding the set of applicable rate limits.
+   * @return set of RateLimitPolicyEntry that are applicable for a stage.
+   */
+  virtual const std::vector<std::reference_wrapper<RateLimitPolicyEntry>>&
```
Comment (mattklein123) [inline include/envoy/router/router_ratelimit.h]: const RateLimitPolicyEntry

File: source/common/http/async_client_impl.cc [hunk 8]
```diff
@@ -3,6 +3,8 @@
 
 namespace Http {
 
+const std::vector<std::reference_wrapper<Router::RateLimitPolicyEntry>>
+    AsyncRequestImpl::NullRateLimitPolicy::rate_limit_policy_entry_ = {};
```
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: '= {}' not needed

File: source/common/router/router_ratelimit.cc [hunk 9]
```diff
@@ -0,0 +1,109 @@
+#include "router_ratelimit.h"
+
+namespace Router {
+
+const std::vector<std::reference_wrapper<RateLimitPolicyEntry>>
+    RateLimitPolicyImpl::empty_rate_limit_ = {};
```
Comment (mattklein123) [inline source/common/router/router_ratelimit.cc]: '= {}' not needed

File: test/mocks/router/mocks.h [hunk 10]
```diff
@@ -38,16 +39,45 @@ class MockRetryState : public RetryState {
 
   DoRetryCallback callback_;
 };
+class TestRateLimitPolicyEntry : public RateLimitPolicyEntry {
```
Comment (mattklein123) [inline test/mocks/router/mocks.h]: newline before this line

File: source/common/router/router_ratelimit.h [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+  // Router::RateLimitPolicy
+  const std::string& routeKey() const override { return route_key_; }
+
+  // Router::RateLimitPolicy
+  const std::vector<std::reference_wrapper<RateLimitPolicyEntry>>&
+  getApplicableRateLimit(int64_t stage = 0) const override;
+
+private:
+  const std::string route_key_;
+  std::vector<std::vector<std::reference_wrapper<RateLimitPolicyEntry>>> rate_limit_entries_;
```
Comment (mattklein123) [inline source/common/router/router_ratelimit.h]: This is broken. You don't store the policies anywhere so this code will crash. You need to actually store the policy implementations somewhere and then also have the reference wrapper list. This probably also means that you don't have 100% coverage on the new code so I would make sure you have 100% coverage. Once that is done I can look again.

File: source/common/router/router_ratelimit.h [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+#include "common/http/filter/ratelimit.h"
+
+namespace Router {
+
+/**
+* Action for service to service rate limiting.
+*/
+class ServiceToServiceAction : public RateLimitAction {
+public:
+  // Router::Action
```
Comment (RomanDzhabarov) [inline source/common/router/router_ratelimit.h]: Router:: RateLimitAction

File: source/common/router/router_ratelimit.h [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+/**
+* Action for request headers rate limiting.
+*/
+class RequestHeadersAction : public RateLimitAction {
+public:
+  RequestHeadersAction(const Json::Object& action)
+      : header_name_(action.getString("header_name")),
+        descriptor_key_(action.getString("descriptor_key")) {}
+
+  // Router::Action
```
Comment (RomanDzhabarov) [inline source/common/router/router_ratelimit.h]: Router::RateLimitAction

File: source/common/router/router_ratelimit.h [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+  const Http::LowerCaseString header_name_;
+  const std::string descriptor_key_;
+};
+
+/**
+ * Action for remote address rate limiting.
+ */
+class RemoteAddressAction : public RateLimitAction {
+public:
+  // Router::Action
```
Comment (RomanDzhabarov) [inline source/common/router/router_ratelimit.h]: ditto

File: test/mocks/router/mocks.h [hunk 15]
```diff
@@ -39,15 +40,46 @@ class MockRetryState : public RetryState {
...
+public:
+  // Router::RateLimitPolicyEntry
+  int64_t stage() const override { return stage_; }
+
+  // Router::RateLimitPolicyEntry
+  const std::string& killSwitchKey() const override { return kill_switch_key_; }
+
+  // Router::RateLimitAction
+  void populateDescriptors(const RouteEntry& route,
```
Comment (RomanDzhabarov) [inline test/mocks/router/mocks.h]: why do we need this and not directly MOCK_CONST_METHOD5?

File: source/common/http/filter/ratelimit.h [hunk 16]
```diff
@@ -10,89 +10,29 @@
...
+
+  const std::string& domain() const { return domain_; }
+  const std::string& localServiceCluster() const { return local_service_cluster_; }
+  int64_t stage() const { return stage_; }
   Runtime::Loader& runtime() { return runtime_; }
   Stats::Store& stats() { return stats_store_; }
 
 private:
   const std::string domain_;
+  int64_t stage_{};
```
Comment (ccaraman) [inline source/common/http/filter/ratelimit.h]: fixed.

File: include/envoy/router/router_ratelimit.h [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+  virtual int64_t stage() const PURE;
+
+  /**
+   * @return runtime key to be set to disable the configuration.
+   */
+  virtual const std::string& killSwitchKey() const PURE;
+};
+
+/**
+ * Rate limiting policy
```
Comment (ccaraman) [inline include/envoy/router/router_ratelimit.h]: added

File: source/common/router/router_ratelimit.h (hunk 12 above)
Comment (ccaraman) [inline source/common/router/router_ratelimit.h]: fixed.

File: source/common/router/router_ratelimit.h (hunk 13 above)
Comment (ccaraman) [inline source/common/router/router_ratelimit.h]: fixed

File: source/common/router/router_ratelimit.h (hunk 14 above)
Comment (ccaraman) [inline source/common/router/router_ratelimit.h]: fixed

File: include/envoy/router/router_ratelimit.h [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+                                   const Http::HeaderMap& headers,
+                                   const std::string& remote_address) const PURE;
+};
+
+typedef std::unique_ptr<RateLimitAction> RateLimitActionPtr;
+
+/**
+ * Rate limit configuration
+ */
+class RateLimitPolicyEntry : public virtual RateLimitAction {
```
Comment (ccaraman) [inline include/envoy/router/router_ratelimit.h]: It wasn't needed. Removed.

File: test/mocks/router/mocks.h (hunk 15 above)
Comment (ccaraman) [inline test/mocks/router/mocks.h]: Fixed in both cases.

File: source/common/router/router_ratelimit.h [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+namespace Router {
+
+/**
+* Action for service to service rate limiting.
+*/
+class ServiceToServiceAction : public RateLimitAction {
+public:
+  // Router::RateLimitAction
+  void populateDescriptors(const Router::RouteEntry& route,
```
Comment (mattklein123) [inline source/common/router/router_ratelimit.h]: nit: all of the function declarations in this file should have full variable names.

File: test/common/http/filter/ratelimit_test.cc [hunk 20]
```diff
@@ -64,6 +30,7 @@ class HttpRateLimitFilterTest : public testing::Test {
         .WillByDefault(Return(true));
     ON_CALL(runtime_.snapshot_, featureEnabled("ratelimit.test_key.http_filter_enabled", 100))
         .WillByDefault(Return(true));
+    ON_CALL(filter_callbacks_, downstreamAddress()).WillByDefault(ReturnRef(EMPTY_STRING));
```
Comment (mattklein123) [inline test/common/http/filter/ratelimit_test.cc]: Per our discussion, move this into the main mock definition so that it has a default.

File: test/mocks/router/mocks.h [hunk 21]
```diff
@@ -39,15 +40,35 @@ class MockRetryState : public RetryState {
   DoRetryCallback callback_;
 };
 
+class TestRateLimitPolicyEntry : public RateLimitPolicyEntry {
```
Comment (mattklein123) [inline test/mocks/router/mocks.h]: If you mock any functions, just mock them all for consistency. Switch this to MockRateLimitPolicyEntry. Any class that has mocks should have constructor/destructor defined in header and implemented in cc file. Same for TestRateLimitPolicy

**Comment by mattklein123** [comment]
rename PR, commit: "rate limit: move HTTP rate limit config into route table"

Have a description that summarizes what this large change does.

File: source/common/router/router_ratelimit.h (hunk 19 above)
Comment (ccaraman) [inline source/common/router/router_ratelimit.h]: fixed.

File: test/common/http/filter/ratelimit_test.cc (hunk 20 above)
Comment (ccaraman) [inline test/common/http/filter/ratelimit_test.cc]: moved to http/mocks.cc

File: test/mocks/http/mocks.h [hunk 22]
```diff
@@ -197,6 +197,7 @@ class MockStreamFilterCallbacksBase {
   Event::MockDispatcher dispatcher_;
   testing::NiceMock<AccessLog::MockRequestInfo> request_info_;
   testing::NiceMock<Router::MockStableRouteTable> route_table_;
+  std::string downstream_address_{};
```
Comment (mattklein123) [inline test/mocks/http/mocks.h]: {} not needed

File: source/common/router/router_ratelimit.h [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+ * Implementation of RateLimitPolicyEntry that holds the action for the configuration.
+ */
+class RateLimitPolicyEntryImpl : public RateLimitPolicyEntry {
+public:
+  RateLimitPolicyEntryImpl(const Json::Object& config);
+
+  // Router::RateLimitPolicyEntry
+  int64_t stage() const override { return stage_; }
+
+  // Router::RateLimitPolicyEntry
```
Comment (mattklein123) [inline source/common/router/router_ratelimit.h]: delete and collapse under previous comment

File: source/common/router/router_ratelimit.h [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+ * Implementation of RateLimitPolicy that reads from the JSON route config.
+ */
+class RateLimitPolicyImpl : public RateLimitPolicy {
+public:
+  RateLimitPolicyImpl(const Json::Object& config);
+
+  // Router::RateLimitPolicy
+  const std::string& routeKey() const override { return route_key_; }
+
+  // Router::RateLimitPolicy
```
Comment (mattklein123) [inline source/common/router/router_ratelimit.h]: delete and collapse

File: test/common/router/router_ratelimit_test.cc [hunk 25]
```diff
@@ -0,0 +1,485 @@
...
+using testing::_;
+using testing::NiceMock;
+using testing::ReturnRef;
+
+namespace Router {
+
+TEST(BadRateLimitConfiguration, MissingActions) {
+  std::string json = R"EOF(
+{
+"virtual_hosts": [
```
Comment (mattklein123) [inline test/common/router/router_ratelimit_test.cc]: proper indent for json

Comment (mattklein123) [inline test/common/router/router_ratelimit_test.cc]: please check all of them

Comment (ccaraman) [inline test/common/router/router_ratelimit_test.cc]: will fix everywhere

File: test/mocks/router/mocks.h [hunk 26]
```diff
@@ -39,16 +40,43 @@ class MockRetryState : public RetryState {
...
-class TestRateLimitPolicy : public RateLimitPolicy {
+class MockRateLimitPolicyEntry : public RateLimitPolicyEntry {
 public:
+  MockRateLimitPolicyEntry();
+  ~MockRateLimitPolicyEntry();
+
+  // Router::RateLimitPolicyEntry
+  MOCK_CONST_METHOD0(stage, int64_t());
+
+  // Router::RateLimitPolicyEntry
```
Comment (mattklein123) [inline test/mocks/router/mocks.h]: collapse under previous comment

File: test/mocks/router/mocks.h [hunk 27]
```diff
@@ -39,16 +40,43 @@ class MockRetryState : public RetryState {
...
+  MockRateLimitPolicy();
+  ~MockRateLimitPolicy();
+
   // Router::RateLimitPolicy
-  bool doGlobalLimiting() const override { return do_global_limiting_; }
+  MOCK_CONST_METHOD0(routeKey, const std::string&());
 
   // Router::RateLimitPolicy
-  const std::string& routeKey() const override { return route_key_; }
+  MOCK_CONST_METHOD1(
```
Comment (mattklein123) [inline test/mocks/router/mocks.h]: collapse

File: test/common/router/router_ratelimit_test.cc [hunk 28]
```diff
@@ -0,0 +1,485 @@
...
+          }
+        ]}
+        ]
+      }
+    ]
+  }
+]
+}
+)EOF";
+  SetUpTest(json);
```
Comment (mattklein123) [inline test/common/router/router_ratelimit_test.cc]: newline before this line

Comment (ccaraman) [inline test/common/router/router_ratelimit_test.cc]: added.

File: source/common/router/router_ratelimit.h [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+  const std::string route_key_;
+};
+
+/**
+ * Action for remote address rate limiting.
+ */
+class RemoteAddressAction : public RateLimitAction {
+public:
+  RemoteAddressAction(const std::string& route_key) : route_key_(route_key) {}
+  // Router::RateLimitAction
```
Comment (mattklein123) [inline source/common/router/router_ratelimit.h]: newline before this line

File: test/common/router/router_ratelimit_test.cc [hunk 30]
```diff
@@ -0,0 +1,509 @@
...
+    {
+      "name": "www2",
+      "domains": ["www.lyft.com"],
+      "routes": [
+        {
+          "prefix": "/",
+          "cluster": "www2",
+          "rate_limits": [
+            {
+              "actions":[ {"type": "bad_type"}]
```
Comment (mattklein123) [inline test/common/router/router_ratelimit_test.cc]: spacing

File: test/common/router/router_ratelimit_test.cc [hunk 31]
```diff
@@ -0,0 +1,509 @@
...
+    {
+      "name": "www2",
+      "domains": ["www.lyft.com"],
+      "routes": [
+        {
+          "prefix": "/",
+          "cluster": "www2",
+          "rate_limits": [
+            {
+              "actions":[ {"type": "request_headers"}]
```
Comment (mattklein123) [inline test/common/router/router_ratelimit_test.cc]: spacing (will stop commenting on this. please check all)

**Comment by mattklein123** [comment]
looks good after nits and docs are added. Please be more careful w/ formatting so I don't have to comment on that stuff so much.

File: docs/configuration/http_conn_man/route_config/rate_limits.rst [hunk 32]
```diff
@@ -0,0 +1,110 @@
...
+  This information is used by the actions if it is installed.
+
+actions
+  *(required, array)* A list of actions that are to be applied for this rate limit configuration.
+  Order matters as the actions are processed sequentially and the descriptors will be composed in
+  that sequence.
+
+.. _config_http_conn_man_route_table_rate_limit_actions:
+
+Actions
```
Comment (ccaraman) [inline docs/configuration/http_conn_man/route_config/rate_limits.rst]: This was copy/pasted from the filter.

File: docs/configuration/http_conn_man/route_config/rate_limits.rst [hunk 33]
```diff
@@ -0,0 +1,110 @@
...
+  *(optional, integer)* Refers to the stage set in the filter. If set, the rate limit configuration
+  only applies to filters with the same stage number and for filters set to default. If not set,
+  the rate limit configuration will apply for all rate limit filters set to default.
+
+kill_switch_key
+  *(optional, string)* The key to be set in runtime to disable this rate limit configuration.
+
+route_key
+  *(optional, string)* Specifies a descriptor value to be used when rate limiting for a route.
+  This information is used by the actions if it is installed.
```
Comment (mattklein123) [inline docs/configuration/http_conn_man/route_config/rate_limits.rst]: "if it is set"

File: docs/configuration/http_conn_man/route_config/route.rst [hunk 34]
```diff
@@ -23,7 +23,8 @@ next (e.g., redirect, forward, rewrite, etc.).
     "rate_limit": "{...}",
```
Comment (mattklein123) [inline docs/configuration/http_conn_man/route_config/route.rst]: delete

File: docs/configuration/http_conn_man/route_config/route_config.rst [hunk 35]
```diff
@@ -55,3 +55,4 @@ response_headers_to_remove
   vcluster
   route_matching
   traffic_shifting
+  rate_limits
```
Comment (mattklein123) [inline docs/configuration/http_conn_man/route_config/route_config.rst]: move right below vcluster in TOC order

File: docs/configuration/http_conn_man/route_config/route_config.rst [hunk 36]
```diff
@@ -52,7 +52,7 @@ response_headers_to_remove
 
   vhost
   route
+  rate_limits
```
Comment (mattklein123) [inline docs/configuration/http_conn_man/route_config/route_config.rst]: Below vcluster (per previous comment...)

Comment (ccaraman) [inline docs/configuration/http_conn_man/route_config/route_config.rst]: :face palm:

**Comment by mattklein123** [comment]
lgtm please get approve from @junr03 or @RomanDzhabarov

File: source/common/router/router_ratelimit.cc [hunk 37]
```diff
@@ -0,0 +1,109 @@
...
+    }
+    rate_limit_entries_.emplace_back(std::move(rate_limit_policy));
+    rate_limit_entries_reference_.emplace_back(rate_limit_policy_reference);
+  }
+}
+
+const std::vector<std::reference_wrapper<const RateLimitPolicyEntry>>&
+    RateLimitPolicyImpl::getApplicableRateLimit(int64_t) const {
+  // Currently return all rate limit policy entries.
+  // TODO: Implement returning only rate limit policy entries that match the stage setting.
```
Comment (junr03) [inline source/common/router/router_ratelimit.cc]: Lets add some sort of disclaimer about this in the docs

File: include/envoy/router/router_ratelimit.h [hunk 38]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * Rate limiting policy.
+ */
+class RateLimitPolicy {
+public:
+  virtual ~RateLimitPolicy() {}
+
+  /**
+   * @param stage the value to for finding applicable rate limit configurations.
```
Comment (ccaraman) [inline include/envoy/router/router_ratelimit.h]: fix
